### PR TITLE
[WIP] Integrare UNUserNotification

### LINKED ISSUE #13033: [Feature Request][PushNotification] Use UNUserNotificationCenter in iOS 10
Currently, calling `PushNotificationIOS.cancelAllLocalNotifications` not remove the notification from the Notification Center

In iOS 10, a new UNUserNotification class was introduced, this class has a method which remove the notifications from notification center, check this `https://developer.apple.com/reference/usernotifications/unusernotificationcenter/1649502-removealldeliverednotifications`

I think that a new methods should be integrate in RN -> `removeAllDeliveredNotifications` and `removeDeliveredNotificationsWithIdentifiers`

### PR DESCRIPTION
**Pull request by gperdomor**
Implementing removeAllDeliveredNotifications and removeDeliveredNotifications for remove notifications from notification center, and getDeliveredNotifications

Thanks for submitting a PR! Please read these instructions carefully:

- [x] Explain the **motivation** for making this change.
- [x] Provide a **test plan** demonstrating that the code is solid.
- [x] Match the **code formatting** of the rest of the codebase.
- [x] Target the `master` branch, NOT a "stable" branch.

## Motivation (required)
Currently, calling PushNotificationIOS.cancelAllLocalNotifications not remove the notification from the Notification Center

In iOS 10, a new UNUserNotification class was introduced, this class has a method which get and remove the notifications from notification center

This PR try to solve that.

## Test Plan (required)
In my case, i'm working with an messaging app, every message is a new notification, when the user tap a notification, the app is opened and the rest of notifications should be gone.

**BEFORE:**
img_1814

**AFTER:**
img_1815

**getDeliveredNotifications response example:**
screen shot 2017-03-21 at 8 22 35 pm

Fix #13033

### CODE REVIEW

**Comment by gperdomor** [comment]
@javache updated :D

**Comment by javache** [comment]
Can you please sign the CLA? Can't merge it without that.

**Comment by gperdomor** [comment]
@javache CLA done

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 1]
```diff
@@ -407,6 +408,30 @@ - (void)handleRegisterUserNotificationSettings:(NSNotification *)notification
...
+    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
+    [center removeDeliveredNotificationsWithIdentifiers:identifiers];
+  }
+}
+
+RCT_EXPORT_METHOD(getDeliveredNotifications:(RCTResponseSenderBlock)callback)
+{
+  if ([UNUserNotificationCenter class]) {
+    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
+    [center getDeliveredNotificationsWithCompletionHandler:callback];
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: This is not going to work. This is typed to be NSArray<UNNotification *>. which can't be send to JS just like that. You'll need to transform it to a dictionary first yourself.

File: Libraries/PushNotificationIOS/PushNotificationIOS.js [hunk 2]
```diff
@@ -165,6 +165,27 @@ class PushNotificationIOS {
...
+   * Remove all delivered notifications from Notification Center
+   */
+  static removeAllDeliveredNotifications() {
+    RCTPushNotificationManager.removeAllDeliveredNotifications();
+  }
+
+  /**
+   * Provides you with a list of the app’s notifications that are still displayed in Notification Center
+   */
+  static getDeliveredNotifications(callback: Function) {
```
Comment (javache) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: Can you add a better type here than function?

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m (hunk 1 above)
Comment (gperdomor) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: Yes, i get this error `RCTJSONStringify() encountered the following error: Invalid type in JSON write (UNNotification)`... Can you help me with that? How can i do that?

Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: You'll need to handle the callback in Objective-C and create an NSArray & NSDictionary with values you actually want to send to JS.

File: Libraries/PushNotificationIOS/PushNotificationIOS.js (hunk 2 above)
Comment (gperdomor) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: @javache please check again :D

Comment (javache) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: I didn't mean adding an invariant, I meant using Flow's function types like this: https://flowtype.org/docs/functions.html

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 3]
```diff
@@ -97,6 +98,31 @@ @implementation RCTPushNotificationManager
   return formattedLocalNotification;
 }
 
+static NSDictionary *RCTFormatUNNotification(UNNotification *notification)
+{
+  NSMutableDictionary *formattedNotification = [NSMutableDictionary dictionary];
+  NSMutableDictionary *formattedContent = [NSMutableDictionary dictionary];
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: This variable is unused

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 4]
```diff
@@ -407,6 +433,39 @@ - (void)handleRegisterUserNotificationSettings:(NSNotification *)notification
...
+    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
+    [center removeDeliveredNotificationsWithIdentifiers:identifiers];
+  }
+}
+
+RCT_EXPORT_METHOD(getDeliveredNotifications:(RCTResponseSenderBlock)callback)
+{
+  if ([UNUserNotificationCenter class]) {
+    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
+    NSMutableArray<NSDictionary *> *formattedNotifications = [NSMutableArray new];
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: Move this inside the callback

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 5]
```diff
@@ -407,6 +433,39 @@ - (void)handleRegisterUserNotificationSettings:(NSNotification *)notification
...
+  
+    [center getDeliveredNotificationsWithCompletionHandler:^(NSArray<UNNotification *> * _Nonnull notifications) {
+      for (UNNotification *notification in notifications) {
+
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: nit: remove empty lines

File: Libraries/PushNotificationIOS/PushNotificationIOS.js [hunk 6]
```diff
@@ -165,6 +165,35 @@ class PushNotificationIOS {
...
+  static removeAllDeliveredNotifications(): void {
+    RCTPushNotificationManager.removeAllDeliveredNotifications();
+  }
+
+  /**
+   * Provides you with a list of the app’s notifications that are still displayed in Notification Center
+   *
+   * @param callback Function which receive an array of delivered notifications
+   */
+  static getDeliveredNotifications(callback: (notifications: [Object]) => void): void {
```
Comment (javache) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: You can type Object here to give more information about the fields you transform https://flowtype.org/docs/objects.html

File: Libraries/PushNotificationIOS/PushNotificationIOS.js [hunk 7]
```diff
@@ -165,6 +165,35 @@ class PushNotificationIOS {
...
+    invariant(
```
Comment (javache) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: No need for this

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 8]
```diff
@@ -407,6 +431,37 @@ - (void)handleRegisterUserNotificationSettings:(NSNotification *)notification
...
+    [center removeDeliveredNotificationsWithIdentifiers:identifiers];
+  }
+}
+
+RCT_EXPORT_METHOD(getDeliveredNotifications:(RCTResponseSenderBlock)callback)
+{
+  if ([UNUserNotificationCenter class]) {
+    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
+    [center getDeliveredNotificationsWithCompletionHandler:^(NSArray<UNNotification *> * _Nonnull notifications) {
+    NSMutableArray<NSDictionary *> *formattedNotifications = [NSMutableArray new];
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: nit: indent this correctly

File: Libraries/PushNotificationIOS/PushNotificationIOS.js (hunk 6 above)
Comment (gperdomor) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: updated

File: Libraries/PushNotificationIOS/PushNotificationIOS.js (hunk 7 above)
Comment (gperdomor) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: updated

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m (hunk 8 above)
Comment (gperdomor) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: fixed

**Comment by LeoNatan** [comment]
I don't understand how exactly was UserNotifications framework "integrated", when at most, the "delivered" API was used. What about handling user notifications from `UNUserNotificationCenterDelegate`? This API is two years old.

**Pull request closed without merging**